platform/engineering/design-docs/2020-05-26-content-validation.md

```diff
@@ -0,0 +1,201 @@
+
+##### Content Sources
+
+Static content currently comes from two sources:
+
+1. The Drupal CMS
+2. The [`vagov-content` repo](https://github.com/department-of-veterans-affairs/vagov-content/blob/master/pages/404.md)
+
+This design doc will focus on validating content from the Drupal CMS.
+
+##### Amazon S3
+
+The frontend of VA.gov is served as static files via Amazon S3. Since `vets-website` currently handles both application and content building, a failure in one blocks both from being deployed to S3 during a full deploy.
+
+###### Broken Link Checking
+
+The [`check-broken-link` middleware](https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/site/stages/build/plugins/check-broken-links/index.js) is a Metalsmith plugin that happens near the end of the Metalsmith static content build pipeline. It does the following:
+
+1. Loop through all of the HTML files in the Metalsmith pipeline
+2. Extract all of the broken `href`/`src` values using the `getBrokenLinks` helper
+3. Formats the result into useful console output
+4. Blocks the deployment on production or logs the error output on lower environments
+
+###### Accessibility Checking
+
+The accessibility checks only happen in CI ([the Integration step in Jenkins](https://github.com/department-of-veterans-affairs/vets-website/blob/master/Jenkinsfile#L81), which only happens during the full `vets-website` build.
+
+Before accessibility checks can be run, the sitemap is generated via the [`create-sitemap` Metalsmith plugin](https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/site/stages/build/plugins/create-sitemaps.js). Nightwatch iterates over all the paths found in the sitemap, and runs the [aXe accessibility checker](https://github.com/department-of-veterans-affairs/vets-website/blob/master/src/platform/testing/e2e/nightwatch-commands/axeCheck.js) to look for violations.
+
+When violations are found, the build breaks and violations must be reported manually.
+
+The accessibility checks are skipped during a content-only deploy.
+
+Additionally, we have the preview server which shows content editors any accessibility violations, but they must deliberately go to the server from a link in Drupal to see this.
+
+#### Pain points
+
+##### Content Writers
+
+Content writers want to quickly draft, publish, and deploy content. However, `vets-website` is only deployed to production once per day, and `vets-website` deploys to staging take 20+ minutes. To allow content writers to move faster, there is a partial deploy, which only includes static page changes (`vagov-content` and Drupal).
+
+Unfortunately, the partial deploys have no content validation. That means that content writers don't have a proactive way to learn if their content includes broken links or accessibility errors.
```

Comment:
Does ` partial deploy` == content deploy?

Someone should definitely confirm broken links don't block content deploys because I thought they did.

Reply:
Oh hey, look at that. So they do.
![image](https://user-images.githubusercontent.com/12970166/83803417-5c5f2800-a661-11ea-98ec-117a8f44769a.png)
From `checkForBrokenLinks` in `jenkins/common.groovy`.

I even made that change, but I totally forgot. My mistake! I've led you astray, @billfienberg.

Reply:
When I say `partial deploy`, I am referring to the content deploy. I got the term `partial deploy` from the [Separate Content Build design doc](https://github.com/department-of-veterans-affairs/va.gov-team/blob/master/platform/engineering/design-docs/2020-04-09-separate-content-build.md#deployment), which links to the [Deploy page of the VA.gov workflow docs](https://department-of-veterans-affairs.github.io/veteran-facing-services-tools/getting-started/workflow/deploy/#partial-deploy--static-page-changes-only).

Reply:
Oh yea. There it is in `common.groovy`.

```
    // Only break the build if broken links are found in master
    if (IS_PROD_BRANCH || contentOnlyBuild) {
      throw new Exception('Broken links found')
    }
  } else {
    echo "Did not find broken links."
  }
```

https://github.com/department-of-veterans-affairs/vets-website/blob/master/jenkins/common.groovy#L177